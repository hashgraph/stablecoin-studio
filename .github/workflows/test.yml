name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test