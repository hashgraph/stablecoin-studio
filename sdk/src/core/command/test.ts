/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
