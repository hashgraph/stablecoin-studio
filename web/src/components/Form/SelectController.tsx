import React, { useEffect, forwardRef } from 'react';
import {
	FormControl,
	FormErrorMessage,
	FormLabel,
	HStack,
	Stack,
	Text,
	useMultiStyleConfig as useChakraMultiStyleConfig,
	Flex,
	type SystemStyleObject,
	Box,
} from '@chakra-ui/react';
import type { ChangeEvent, ReactNode, Ref } from 'react';
import { Controller } from 'react-hook-form';
import type { Props as ReactSelectProps } from 'react-select';
import type { Control, UseControllerProps } from 'react-hook-form';

import {
	Select as ChakraSelect,
	type SelectComponentsConfig,
	type GroupBase,
	type SelectInstance,
} from 'chakra-react-select';
import Icon from '../Icon';
import { merge as _merge } from 'lodash';
import type { Variant } from 'chakra-react-select/dist/types/types';

export interface SelectOption {
	value: string | number;
	label: string | number;
}

export const partsListByTheme: Array<
	| 'wrapper'
	| 'container'
	| 'addonLeft'
	| 'addonRight'
	| 'addonError'
	| 'addonDown'
	| 'menuList'
	| 'label'
	| 'valueSelected'
	| 'option'
	| 'optionSelected'
> = [
	'wrapper',
	'container',
	'addonLeft',
	'addonRight',
	'addonError',
	'addonDown',
	'menuList',
	'label',
	'valueSelected',
	'option',
	'optionSelected',
];

type Parts = typeof partsListByTheme;

export type SelectThemeStyle = Record<Parts[number], SystemStyleObject>;

export type Option = { label: string | ReactNode; value: number | string };
export interface SelectControllerProps {
	name: string;
	rules?: UseControllerProps['rules'];
	label?: string | null;
	placeholder?: string | null;
	options: Array<Option>;
	control: Control<Record<string, SelectOption['value']>>;
	isRequired?: boolean;
	isSearchable?: boolean;
	isDisabled?: boolean;
	onChangeAux?: React.ChangeEventHandler<HTMLInputElement>;
	onBlurAux?: ReactSelectProps['onBlur'];
	showErrors?: boolean;
	labelProps?: Object;
	'data-testid'?: string;
	onMenuOpen?: () => void;
	onMenuClose?: () => void;
	defaultValue?: SelectOption['value'];
	addonDown?: ReactNode;
	addonRight?: ReactNode;
	addonError?: ReactNode;
	addonLeft?: ReactNode;
	variant?: Variant;
	size?: ReactSelectProps['size'];
	overrideStyles?: Partial<SelectThemeStyle>;
	noOptionsMessage?: ReactSelectProps['noOptionsMessage'];
	isMulti?: ReactSelectProps['isMulti'];
}

export type SelectConfigProps = {
	variant: Record<string, SystemStyleObject>;
	isInvalid?: boolean;
	isDisabled?: boolean;
	addonDown?: ReactNode;
	addonError?: ReactNode;
	addonLeft?: ReactNode;
	addonRight?: ReactNode;
	hasValue?: boolean;
};

const IconContainer = ({ children, sx }: { children: ReactNode; sx?: SystemStyleObject }) => (
	<Flex alignItems='center' justifyContent='center' sx={sx}>
		{children}
	</Flex>
);

export const useStyles = ({
	variant,
	size,
	isInvalid,
	isDisabled,
	addonLeft,
	addonRight,
	addonError,
	addonDown,
	overrideStyles,
}: {
	variant: Variant;
	size: ReactSelectProps['size'];
	isInvalid: boolean;
	isDisabled: boolean;
	addonDown: ReactNode;
	addonRight: ReactNode;
	addonError: ReactNode;
	addonLeft: ReactNode;
	overrideStyles?: Partial<SelectThemeStyle>;
}) => {
	const themeStyles = useChakraMultiStyleConfig('Select', {
		variant,
		size,
		isInvalid,
		isDisabled,
		addonLeft,
		addonRight,
		addonError,
		addonDown,
	});
	const styles = React.useMemo(
		() => _merge({}, themeStyles, overrideStyles),
		[themeStyles, overrideStyles],
	);
	return styles as SelectThemeStyle;
};

export const useComponents = ({
	variant,
	addonLeft,
	addonRight,
	addonError,
	addonDown,
	size,
	isInvalid,
	isDisabled,
	placeholder,
	styles,
}: {
	variant: Variant;
	addonLeft?: ReactNode;
	addonRight?: ReactNode;
	addonError?: ReactNode;
	addonDown?: ReactNode;
	size?: ReactSelectProps['size'];
	isInvalid?: boolean;
	isDisabled?: boolean;
	placeholder?: string | ReactNode;
	styles: SelectThemeStyle;
}): SelectComponentsConfig<unknown, boolean, GroupBase<unknown>> => {
	return {
		Placeholder: () => null,
		IndicatorSeparator: () => null,
		DropdownIndicator: () => <IconContainer sx={styles.addonDown}>{addonDown}</IconContainer>,
		SelectContainer: ({ children }) => {
			return (
				<Box sx={styles.wrapper}>
					{addonLeft && <IconContainer sx={styles.addonLeft}>{addonLeft}</IconContainer>}
					{children}
					{addonRight && !isInvalid && (
						<IconContainer sx={styles.addonRight}>{addonRight}</IconContainer>
					)}
					{isInvalid && <IconContainer sx={styles.addonError}>{addonError}</IconContainer>}
				</Box>
			);
		},
		ValueContainer: ({ children, hasValue, selectProps: { inputValue } }) => {
			const _styles = useChakraMultiStyleConfig('Select', {
				variant,
				size,
				isInvalid,
				isDisabled,
				hasValue: hasValue || inputValue,
			}) as SelectThemeStyle;

			return (
				<Box sx={_styles.container}>
					<Text as='span' data-testid='select-placeholder' sx={_styles.label}>
						{placeholder}
					</Text>
					{children}
				</Box>
			);
		},
		SingleValue: ({ children }) => <Box sx={styles.valueSelected}>{children}</Box>,
		MenuList: ({ children }) => <Box sx={styles.menuList}>{children}</Box>,
	};
};

const SelectController = forwardRef(
	(
		{
			name,
			rules,
			label,
			placeholder,
			options,
			control,
			isRequired = false,
			isDisabled = false,
			onChangeAux,
			onBlurAux,
			showErrors = true,
			'data-testid': dataTestId = name,
			size,
			variant = 'outline',
			addonLeft,
			addonRight,
			addonError,
			addonDown = <Icon name='CaretDown' />,
			labelProps,
			overrideStyles,
			defaultValue,
			...props
		}: SelectControllerProps,
		ref: Ref<SelectInstance<unknown, boolean, GroupBase<unknown>>>,
	) => {
		const styles = useStyles({
			variant,
			addonRight,
			addonError,
			addonDown,
			addonLeft,
			size,
			isInvalid: false,
			isDisabled,
			overrideStyles,
		});
		const components = useComponents({
			addonLeft,
			addonRight,
			addonError,
			addonDown,
			placeholder,
			isInvalid: false,
			isDisabled,
			styles,
			variant,
		});

		return (
			<Controller
				control={control}
				name={name}
				defaultValue={defaultValue}
				rules={rules}
				render={({ field: { onChange, value }, fieldState: { invalid, error } }) => {
					useEffect(() => {
						if (defaultValue) {
							const index = parseInt(defaultValue as string);
							const defaultOption = options[index];

							onChange(defaultOption);
						}
					}, []);
					const onChangeCustom = (event: ChangeEvent<HTMLInputElement>) => {
						onChange(event);
						onChangeAux && onChangeAux(event);
					};

					const onBlurCustom = (event: React.FocusEvent<HTMLInputElement>) => {
						onBlurAux && onBlurAux(event);
					};

					return (
						<Stack w='full'>
							<FormControl data-testid='form_control' isInvalid={invalid}>
								{label && (
									<FormLabel {...labelProps}>
										<HStack>
											<Text data-testid='selector-label'>{label}</Text>
											{isRequired && <Text color='red'>*</Text>}
										</HStack>
									</FormLabel>
								)}

								<ChakraSelect
									isInvalid={invalid}
									name={name}
									options={options}
									onChange={onChangeCustom as ReactSelectProps['onChange']}
									onBlur={onBlurCustom as ReactSelectProps['onBlur']}
									placeholder={placeholder}
									value={value}
									components={components}
									ref={ref}
									isDisabled={isDisabled}
									chakraStyles={{
										option: (_, state) => ({
											// option needs to be styled like this, otherwise it doesn't let user select the options
											...styles.option,
											...(state.isSelected && styles.optionSelected),
										}),
									}}
									variant={variant}
									{...props}
								/>
								{showErrors && (
									<FormErrorMessage data-testid='form-error-msg'>
										{error && error.message}
									</FormErrorMessage>
								)}
							</FormControl>
						</Stack>
					);
				}}
			/>
		);
	},
);

SelectController.displayName = 'SelectController';

export { SelectController };
