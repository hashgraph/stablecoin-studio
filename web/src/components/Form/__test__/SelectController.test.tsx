import { SelectController } from '../SelectController';
import type {
	SelectControllerProps as BaseSelectControllerProps,
	SelectOption,
} from '../SelectController';
import { render } from '../../../test/index';
import type { UseFormReturn } from 'react-hook-form';
import { useForm } from 'react-hook-form';
import { act, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import translations from '../../../translations/en/global.json';

describe(`<${SelectController.name} />`, () => {
	type SelectControllerProps = Omit<BaseSelectControllerProps, 'control'>;

	const defaultProps = {
		name: 'selection',
		options: [
			{ value: 'value1', label: 'Value 1' },
			{ value: 'value2', label: 'Value 2' },
			{ value: 'value3', label: 'Value 3' },
		],
		label: 'Select an option',
		rules: {},
	};

	let form: UseFormReturn;

	const RenderWithForm = (props: Omit<SelectControllerProps, 'control'>) => {
		const localForm = useForm({
			mode: 'onChange',
			criteriaMode: 'all',
		});

		form = localForm;

		return <SelectController control={localForm.control} {...props} />;
	};

	const factoryComponent = (props: SelectControllerProps = defaultProps) => {
		return render(<RenderWithForm {...props} />);
	};

	test('reports correctly the values', () => {
		const component = factoryComponent();

		expect(form.getValues()[defaultProps.name]).toBeUndefined();

		const selector = component.getByRole('combobox');
		expect(selector).toBeInTheDocument();
		userEvent.click(selector);

		const optionToSelect = component.getByText(defaultProps.options[1].label);
		userEvent.click(optionToSelect);

		expect(form.getValues()[defaultProps.name]).toEqual(defaultProps.options[1]);
	});

	test('shows selected value provided by defaultValue', () => {
		const indexOption = 1;
		const defaultValue = defaultProps.options[indexOption];
		const component = factoryComponent({ ...defaultProps, defaultValue: indexOption.toString() });

		expect(component.queryByText(defaultProps.options[0].label)).not.toBeInTheDocument();
		expect(component.queryByText(defaultProps.options[1].label)).toBeInTheDocument();
		expect(component.queryByText(defaultProps.options[2].label)).not.toBeInTheDocument();

		expect(form.getValues()[defaultProps.name]).toEqual(defaultValue);
	});

	test('incorporates rules to the controller', async () => {
		const notSelectableOption = defaultProps.options[1];
		const component = factoryComponent({
			...defaultProps,
			showErrors: true,
			rules: {
				validate: {
					blacklist: (value: SelectOption) => value.value !== notSelectableOption.value,
				},
			},
		});

		let errorMessage = component.queryByTestId('form-error-msg');
		expect(errorMessage).not.toBeInTheDocument();

		const selector = component.getByRole('combobox');
		expect(selector).toBeInTheDocument();
		userEvent.click(selector);

		const optionToSelect = component.getByText(notSelectableOption.label);
		userEvent.click(optionToSelect);

		await waitFor(() => {
			errorMessage = component.getByTestId('form-error-msg');
			expect(errorMessage).toBeInTheDocument();
		});
	});

	test('shows placeholder with default value', () => {
		const component = factoryComponent();

		const placeholder = component.getByTestId('select-placeholder');
		expect(placeholder).toBeEmptyDOMElement();

		expect(component.asFragment()).toMatchSnapshot('DefaultPlaceholder');
	});

	test('shows placeholder with the provided value', () => {
		const placeholderText = 'Value of placeholder';
		const component = factoryComponent({
			...defaultProps,
			placeholder: placeholderText,
		});

		const placeholder = component.getByTestId('select-placeholder');
		expect(placeholder).toHaveTextContent(placeholderText);

		expect(component.asFragment()).toMatchSnapshot('CustomPlaceholder');
	});

	test('shows error with showErrors=true', () => {
		const component = factoryComponent({
			...defaultProps,
			showErrors: true,
		});

		act(() => {
			form.setError(defaultProps.name, { type: 'expired' });
		});

		expect(component.getByTestId('form-error-msg')).toBeInTheDocument();
		expect(component.asFragment()).toMatchSnapshot('ShowingError');
	});

	test("doesn't show error with showError=false", () => {
		const component = factoryComponent({
			...defaultProps,
			showErrors: false,
		});

		act(() => {
			form.setError(defaultProps.name, { type: 'expired' });
		});

		expect(component.queryByTestId('form-error-msg')).toBeNull();
		expect(component.asFragment()).toMatchSnapshot('NotShowingError');
	});

	test("doesn't require field with isRequired=false", async () => {
		const component = factoryComponent({
			...defaultProps,
			showErrors: true,
		});

		expect(form.getValues()[defaultProps.name]).toBeUndefined();

		act(() => {
			form.setValue(defaultProps.name, '', {
				shouldValidate: true,
				shouldDirty: true,
				shouldTouch: true,
			});
		});

		await waitFor(() => {
			expect(component.queryByTestId('form-error-msg')).not.toBeInTheDocument();
		});
	});

	test('requires field with isRequired=true', async () => {
		const required = translations.validations.required;
		const component = factoryComponent({
			...defaultProps,
			showErrors: true,
			rules: {
				required,
			},
		});

		act(() => {
			form.setValue(defaultProps.name, '', {
				shouldValidate: true,
				shouldDirty: true,
				shouldTouch: true,
			});
		});

		await waitFor(() => {
			expect(component.getByTestId('form-error-msg')).toBeInTheDocument();
		});
	});

	test('propagates prop onChange to select', async () => {
		const onChange = jest.fn();
		const component = factoryComponent({
			...defaultProps,
			onChangeAux: onChange,
		});

		const selector = component.getByRole('combobox');
		expect(selector).toBeInTheDocument();
		userEvent.click(selector);

		const optionToSelect = component.getByText(defaultProps.options[1].label);
		userEvent.click(optionToSelect);

		expect(onChange).toHaveBeenCalled();
	});

	test('propagates prop onBlur to select', async () => {
		const onBlur = jest.fn();
		const component = factoryComponent({
			...defaultProps,
			onBlurAux: onBlur,
		});

		const selector = component.getByRole('combobox');
		expect(selector).toBeInTheDocument();
		userEvent.click(selector);
		userEvent.click(document.body);

		expect(onBlur).toHaveBeenCalled();
	});

	test('shows label', async () => {
		const textLabel = 'LabelForSelect';
		const component = factoryComponent({
			...defaultProps,
			label: textLabel,
		});

		const label = component.getByTestId('selector-label');
		expect(label).toBeInTheDocument();
		expect(label).toHaveTextContent(textLabel);
	});

	test('dont show label when label is empty', async () => {
		const component = factoryComponent({
			...defaultProps,
			label: '',
		});

		const label = component.queryByTestId('selector-label');
		expect(label).not.toBeInTheDocument();
	});

	test('cannot be used when is disabled', () => {
		const onChange = jest.fn();
		const placeholderText = 'disabled selector';
		const component = factoryComponent({
			...defaultProps,
			isDisabled: true,
			onChangeAux: onChange,
			placeholder: placeholderText,
		});

		const selector = component.queryByRole('combobox');
		expect(selector).not.toBeInTheDocument();

		const placeholder = component.getByText(placeholderText);
		userEvent.click(placeholder);

		defaultProps.options.forEach((option) => {
			expect(component.queryByText(option.label)).not.toBeInTheDocument();
		});

		expect(onChange).not.toHaveBeenCalled();
	});
});
