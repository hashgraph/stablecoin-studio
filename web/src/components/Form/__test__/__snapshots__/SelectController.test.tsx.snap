// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectController /> doesn't show error with showError=false: NotShowingError 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      data-testid="form_control"
      role="group"
    >
      <label
        class="chakra-form__label css-1jbsrib"
        data-invalid=""
        for="field-:r6:"
        id="field-:r6:-label"
      >
        <div
          class="chakra-stack css-84zodg"
        >
          <p
            class="chakra-text css-mgupmj"
            data-testid="selector-label"
          >
            Select an option
          </p>
        </div>
      </label>
      <div
        class="css-hfeh7c"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-8-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-1gc7io2"
          data-invalid="true"
        >
          <div
            class="css-3y2vuh"
          >
            <span
              class="chakra-text css-19fpr04"
              data-testid="select-placeholder"
            />
            <div
              class=" css-svnjlw"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-8-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                aria-invalid="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=" css-146k1ar-input"
                id="field-:r6:"
                role="combobox"
                spellcheck="false"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-hfbj6y"
          >
            <div
              class="css-1aabyhj"
            >
              <svg
                class="chakra-icon css-13otjrl"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="0 0 256 256"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="none"
                  height="256"
                  width="256"
                />
                <polyline
                  fill="none"
                  points="208 96 128 176 48 96"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="16"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="selection"
          type="hidden"
          value=""
        />
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<SelectController /> shows error with showErrors=true: ShowingError 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      data-testid="form_control"
      role="group"
    >
      <label
        class="chakra-form__label css-1jbsrib"
        data-invalid=""
        for="field-:r5:"
        id="field-:r5:-label"
      >
        <div
          class="chakra-stack css-84zodg"
        >
          <p
            class="chakra-text css-mgupmj"
            data-testid="selector-label"
          >
            Select an option
          </p>
        </div>
      </label>
      <div
        class="css-hfeh7c"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-7-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-1gc7io2"
          data-invalid="true"
        >
          <div
            class="css-3y2vuh"
          >
            <span
              class="chakra-text css-19fpr04"
              data-testid="select-placeholder"
            />
            <div
              class=" css-svnjlw"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-7-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                aria-invalid="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=" css-146k1ar-input"
                id="field-:r5:"
                role="combobox"
                spellcheck="false"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-hfbj6y"
          >
            <div
              class="css-1aabyhj"
            >
              <svg
                class="chakra-icon css-13otjrl"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="0 0 256 256"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="none"
                  height="256"
                  width="256"
                />
                <polyline
                  fill="none"
                  points="208 96 128 176 48 96"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="16"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="selection"
          type="hidden"
          value=""
        />
      </div>
      <div
        aria-live="polite"
        class="chakra-form__error-message css-2e3jkk"
        data-testid="form-error-msg"
        id="field-:r5:-feedback"
      />
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<SelectController /> shows placeholder with default value: DefaultPlaceholder 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      data-testid="form_control"
      role="group"
    >
      <label
        class="chakra-form__label css-1jbsrib"
        for="field-:r3:"
        id="field-:r3:-label"
      >
        <div
          class="chakra-stack css-84zodg"
        >
          <p
            class="chakra-text css-mgupmj"
            data-testid="selector-label"
          >
            Select an option
          </p>
        </div>
      </label>
      <div
        class="css-hfeh7c"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-5-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-1gc7io2"
        >
          <div
            class="css-3y2vuh"
          >
            <span
              class="chakra-text css-19fpr04"
              data-testid="select-placeholder"
            />
            <div
              class=" css-svnjlw"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-5-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=" css-146k1ar-input"
                id="field-:r3:"
                role="combobox"
                spellcheck="false"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-hfbj6y"
          >
            <div
              class="css-1aabyhj"
            >
              <svg
                class="chakra-icon css-13otjrl"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="0 0 256 256"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="none"
                  height="256"
                  width="256"
                />
                <polyline
                  fill="none"
                  points="208 96 128 176 48 96"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="16"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="selection"
          type="hidden"
          value=""
        />
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<SelectController /> shows placeholder with the provided value: CustomPlaceholder 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1h0sumw"
  >
    <div
      class="chakra-form-control css-1kxonj9"
      data-testid="form_control"
      role="group"
    >
      <label
        class="chakra-form__label css-1jbsrib"
        for="field-:r4:"
        id="field-:r4:-label"
      >
        <div
          class="chakra-stack css-84zodg"
        >
          <p
            class="chakra-text css-mgupmj"
            data-testid="selector-label"
          >
            Select an option
          </p>
        </div>
      </label>
      <div
        class="css-hfeh7c"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-6-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-1gc7io2"
        >
          <div
            class="css-3y2vuh"
          >
            <span
              class="chakra-text css-19fpr04"
              data-testid="select-placeholder"
            >
              Value of placeholder
            </span>
            <div
              class=" css-svnjlw"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-6-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=" css-146k1ar-input"
                id="field-:r4:"
                role="combobox"
                spellcheck="false"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-hfbj6y"
          >
            <div
              class="css-1aabyhj"
            >
              <svg
                class="chakra-icon css-13otjrl"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="0 0 256 256"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="none"
                  height="256"
                  width="256"
                />
                <polyline
                  fill="none"
                  points="208 96 128 176 48 96"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="16"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="selection"
          type="hidden"
          value=""
        />
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
